### ISSUE TITLE
Use an optimized malloc replacement for better CGAL performance

### ISSUE THREAD

**Issue by thehans** [openscad/openscad#3930]
It was recently brought up in IRC by @ChrisCoxArt that profiling showed a majority of render time spent doing mallocs for CGAL / GMP.  I've also noticed this excessive malloc/free usage from my profiling in the past, and decided to look at alternatives today.

After some brief searching, I came across this project and decided to give it a try:  https://github.com/microsoft/mimalloc

I downloaded and built from source, then used their [dynamic override](https://github.com/microsoft/mimalloc#override-on-linux-bsd) method on Linux 
```
env LD_PRELOAD=/path/to/libmimalloc.so openscad-nightly
```

Initial results showed a promising ~35% speedup of CGAL render, so I decided to test some of the other malloc replacements.

For each test, I opened `example024.scad` in GUI, previewed, then rendered and recorded render time, along with memory sizes from `gnome-system-monitor`.

|                                                           | Time   | Resident | Memory | Shared |  Virtual  |
| --------------------------------------------------------- | ------ | -------- | ------ | ------ | --------- |
| default                                                   | 28.272 |  464.5   | 349.7  | 115.0  |   3.3 GiB |
| [jemalloc](https://github.com/jemalloc/jemalloc)          | 23.654 |  386.2   | 270.4  | 115.8  |   1.0 GiB |
| [Mesh](https://github.com/plasma-umass/Mesh)<sup>1</sup>  | 22.316 |  408.8   |  14.8  | 394.1  |  66.8 GiB |
| [tcmalloc](https://github.com/google/tcmalloc)            | 18.979 |  439.2   | 321.3  | 117.9  | 896.6 MiB |
| [mimalloc](https://github.com/microsoft/mimalloc)         | 18.273 |  466.7   | 351.2  | 115.5  |   1.1 GiB |
| [Hoard](https://github.com/emeryberger/Hoard)<sup>2</sup> |    DNF |    DNF   |   DNF  |   DNF  |   DNF     |
| [ptmalloc3](http://www.malloc.de/en/)<sup>2</sup>         |    DNF |    DNF   |   DNF  |   DNF  |   DNF     |

1. Segfault on app close
    ```
    segfault (128/(nil)): in arena? 0
    Aborted (core dumped)
    ```
2. Segfault on app start
    ```
    Segmentation fault (core dumped)
    ```

mimalloc was the most time efficient for this test case, with tcmalloc a close second.

Test Machine:
Linux Mint 20.2
AMD Ryzen 3900X CPU
libc-2.31.so

**Comment by ChrisCoxArt** [comment]
I don't mean our own malloc replacement - just something that would work for the known cases in CGAL/GMP (thus minimizing the risk area, and thus the testing required).  But I'm still learning the code paths and options available.

**Comment by thehans** [comment]
@kintel I've noticed you working on some mac related fixes recently, would you be willing to try a debug build of #3943 and get a stacktrace?  
Or if you have any other ideas/input on how to get mimalloc running on mac (info about interpose, zone options etc.), it would be appreciated.

I'm eager to get this merged since I feel like it would be a nice overall performance boost for us, but I'm pretty much stuck without someone to assist with mac-specific issues (openscad apparently crashes on every test, see previous comment).

I suppose another option would be to just enable mimalloc on all platforms *except* mac, until we can get that sorted.

**Comment by kintel** [comment]
I'll take a look.

**Comment by kintel** [comment]
It's weird; to test I just added an `exit(0)` in OpenSCAD directly after parsing the cmd-line args, and that causes the binary to become flaky: It sometimes succeeds (75% of the time), and sometimes crashes in `mi_free_generic()` and sometimes reports free of unallocated memory.

I don't have any great ideas, except continuing to test different permutations of functionality until I can reliably break stuff.

**Comment by thehans** [comment]
OK, based on that I'm guessing that maybe some memory is being allocated before the `mi_*` overrides become active (assuming that's possible?).  Then later when free gets called, `mi_free_generic` has no record of that memory being allocated.

I haven't looked into mimalloc source to see how those get registered exactly, or if there's anything we can do to ensure it happens before any allocations.

I decided to try disabling `MI_OVERRIDE` so malloc is not globally overridden, and instead register mimalloc functions directly with GMP, CGAL allocator, and c++ new/delete.

I think this should give 99% of the benefits, while hopefully avoiding some of the trickier issues.
And from the test server results, it runs fine on Mac... but now Windows seems to be the one freaking out.
I have a windows machine I can do some testing on at least, so I'll try to sort that out in the next few days.

@kintel I would still be interested in seeing if Mac gets similar performance boost from this change, if you want to do some benching.

**Comment by kintel** [comment]
Cool, I'll run some tests.

Do you have a nice way of running performance tests on the cmd-line? I was thinking zsh+time+TIMEFMT to record both performance and memory usage, but haven't looked into the details yet

**Comment by thehans** [comment]
I've mostly just been testing from GUI and noting the render time from openscad's Console, with a fresh process each time.  `time` is definitely useful if using cmdline though.  
I also have only been trying out a couple tests like example024 (used in my original post above) and example006.

**Comment by kintel** [comment]
A casual test shows a speedup of ~1.8x for example024.
Both tests run on a M1 MacBook Air running under Rosetta2.

**Env**
```
TIMEFMT=$'%U user  %S  system  %P  cpu  %*E total\n%M kB max memory'
```

**master**
```
$ time ./OpenSCAD.app/Contents/MacOS/OpenSCAD ../examples/Old/example024.scad -o out.stl
128.50s user  0.20s  system  99%  cpu  2:09.34 total
1318512 kB max memory
```

**thehans:malloc**
```
$ time ./OpenSCAD.app/Contents/MacOS/OpenSCAD ../examples/Old/example024.scad -o out.stl
72.09s user  0.15s  system  99%  cpu  1:12.63 total
1182144 kB max memory
```

**Comment by thehans** [comment]
Very nice to see the relative speed boost for is even better on mac!

**Comment by kintel** [comment]
I don't know if Rosetta has any extra penalty on malloc ops. I'm working (slowly) towards a universal build, so I'll know more once that runs.

**Comment by ChrisCoxArt** [comment]
> @ChrisCoxArt, were you able to complete a local build on mac? I know you had some issues previously but not sure if they were all resolved. Do you have access to x86 in addition to M1 machines?

Ooops, didn't see this until now for some reason.
Yes, I have x86 and M1 based Mac systems. Yes, I got builds working on both (M1 was a bit of a hack though - Kintel seems to be working on the same sort of changes and making the hard decisions that I outlined).

**Comment by kintel** [comment]
Another casual test on both Rosetta and M1 (on a different M1 machine this time). Same env as above.
Not sure why rosetta is so much faster on this machine, but my test is very casual.

**Rosetta master**
```
93.83s user  0.44s  system  99%  cpu  1:34.69 total
383896 kB max memory
```
**Rosetta thehans:malloc**
```
71.45s user  0.37s  system  98%  cpu  1:13.21 total
299524 kB max memory
```

**M1 master**
```
61.53s user  0.49s  system  98%  cpu  1:03.03 total
375616 kB max memory
```

**M1 thehans:malloc**
```
47.34s user  0.12s  system  98%  cpu  48.195 total
303104 kB max memory
```

**Comment by thehans** [comment]
Closed by #3943
